rubix-bookkeeper/src/test/java/com/qubole/rubix/bookkeeper/TestWorkerBookKeeper.java

```diff
@@ -0,0 +1,134 @@
+ * limitations under the License. See accompanying LICENSE file.
+ */
+
+package com.qubole.rubix.bookkeeper;
+
+import com.codahale.metrics.MetricRegistry;
+import com.qubole.rubix.spi.BookKeeperFactory;
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.RetryingBookkeeperClient;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.thrift.shaded.transport.TSocket;
+import org.apache.thrift.shaded.transport.TTransportException;
+import org.mockito.ArgumentMatchers;
+import org.testng.annotations.Test;
+
+import java.io.FileNotFoundException;
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestWorkerBookKeeper
+{
+  private static final Log log = LogFactory.getLog(TestBookKeeperServer.class.getName());
+
+  private final Configuration conf = new Configuration();
+
+  /**
+   * Verify that WorkerBookKeeper throws the correct exception when asked to handle heartbeats.
+   *
+   * @throws FileNotFoundException if the parent directory for the cache cannot be found when initializing the BookKeeper.
+   * @throws InterruptedException if the current thread is interrupted while starting the BookKeeper server.
+   */
+  @Test(expectedExceptions = UnsupportedOperationException.class)
+  public void testHandleHeartbeat() throws FileNotFoundException, InterruptedException
+  {
+    CacheConfig.setServiceRetryInterval(conf, 2000);
+    CacheConfig.setOnMaster(conf, true);
+
+    startBookKeeperServer();
+    final WorkerBookKeeper workerBookKeeper = new WorkerBookKeeper(conf, new MetricRegistry());
+    workerBookKeeper.handleHeartbeat("");
+  }
+
+  /**
+   * Verify that the heartbeat service correctly makes a connection using a BookKeeper client.
+   *
+   * @throws TTransportException if the BookKeeper client cannot be created.
+   * @throws InterruptedException if the current thread is interrupted while starting the BookKeeper server.
+   */
+  @Test
+  public void testHeartbeatRetryLogic_noRetriesNeeded() throws TTransportException, InterruptedException
+  {
+    final BookKeeperFactory bookKeeperFactory = mock(BookKeeperFactory.class);
+    CacheConfig.setServiceRetryInterval(conf, 2000);
+    CacheConfig.setOnMaster(conf, true);
+
+    when(bookKeeperFactory.createBookKeeperClient(anyString(), ArgumentMatchers.<Configuration>any())).thenReturn(
+        new RetryingBookkeeperClient(
+            new TSocket("localhost", CacheConfig.getServerPort(conf), CacheConfig.getClientTimeout(conf)),
+            CacheConfig.getMaxRetries(conf)));
+
+    startBookKeeperServer();
+    final WorkerBookKeeper.HeartbeatService heartbeatService = new WorkerBookKeeper.HeartbeatService(conf, bookKeeperFactory);
+  }
+
+  /**
+   * Verify that the heartbeat service no longer attempts to connect once it runs out of retry attempts.
```

Comment:
I am not sure if its possible with mock in Java. Can you test if connection is established after a few failures ? Skip if its too hard.

Reply:
Added unit test for connecting after a few retries (see line 99 `testHeartbeatRetryLogic_connectAfterRetries()`)